src/utp_stream.cpp

```diff
@@ -2468,8 +2505,22 @@ void utp_socket_impl::ack_packet(packet* p, time_point const& receive_time
 
 	m_rtt.add_sample(rtt / 1000);
 	if (rtt < min_rtt) min_rtt = rtt;
-	// tune receive buffer size according to the new RTT value
-	m_in_buf_size = adjust_receive_buf_size(m_rtt.mean()*1000);
+
+    // update the all-time mininimum RTT value
+	if (m_min_rtt < min_rtt)
+    {
+        m_min_rtt = min_rtt;
+        // and tune receive buffer size according to it
+        boost::int32_t const new_capacity = get_optimized_receive_buf_capacity(m_min_rtt/1000);
+        // only use surplus mechanism if capacity should really
+        // descrease. In case it should increse, like after first RTT,
+        // change it directly;
+        if (new_capacity <= m_receive_buffer_capacity - m_receive_buffer_surplus)
+            m_receive_buffer_surplus += (m_receive_buffer_capacity - m_receive_buffer_surplus - new_capacity);
+        else
+            m_receive_buffer_capacity = new_capacity;
```

Comment:
I don't understand this logic. as far as I can tell, this would still make is pull receive window "from under the feet" of the sender. If we have advertised 1 MiB, the sender may already have sent that megabyte. The only way the receive window is "replenished" is by our upper layer consuming bytes, in which case the advertised receive window is increased. This is the only opportunity we have to decrease the window and preserve consistency.

Reply:
I think at this point in the code, you should either adjust the ``surplus`` variable (in any direction) *OR* increase ``m_receive_buffer_capacity``.

Reply:
The problem is, we can't start with the buffer of maximal possible capacity (for the sake of not blowing away Raspberry Pi's with lots of saved torrents), and only decrease it afterward. We must start with 1 MByte buffer, _increase_ it after the first round trip, and only then start shrinking it on further `m_min_rtt` updates.
This construction enables us to handle the increase in buffer capacity after first `m_min_rtt` update. As it is written now, it does not account for all possible sequences of increase and decrease of buffer capacity. But it does not have to because the only possible scenario is:
`default -> increase -> decrease -> decrease ...`,

When the buffer capacity must be decreased, we first look if the new capacity is less than the capacity we had already _planned_ our buffer to have (`m_receive_buffer_capacity - m_receive_buffer_surplus`). If it is, we add the difference between our current _planned_ capacity and our new _planned_ capacity to `m_receive_buffer_surplus`.